### ISSUE TITLE
BatchNormalization in mode=2 Question

### ISSUE THREAD

**Issue by konstantinberlin** [fchollet/keras#3085]
Related to #2944 

I am trying to understand the current BatchNormalization code after the introduction of mode=2, when using mode=2. Is it correct that during the prediction phase the prediction is made using the aggregate statistic of the given batch (so mean and std is computed from prediction batch). So if, lets say, I make a prediction using a large batch size, it would be one value, but if I make a forward prediction on a single value, it could swing wildly to a different value?

Aka prediction is stochastic, rather than deterministic.

Please make sure that the boxes below are checked before you submit your issue. Thank you!

- [X] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps

- [X] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps

- [X] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

**Comment by fchollet** [comment]
Yes, that's correct.

**Comment by konstantinberlin** [comment]
You don't that this is very dangerous behavior? Imagine someone trains their model, validates it using batches, deploys it in a situation where the prediction is done one element at a time. There are very few situations where repeated call to prediction giving different answers is ok.

**Comment by fchollet** [comment]
It's a useful behavior, and its description is perfectly clear and unambiguous. People who need it will use it.
